### PR TITLE
WIP: 13 add erc20 tests into ci

### PR DESCRIPTION
**Pull request by vasiliy-zaznobin**

### CODE REVIEW

File: evm_loader/ERC20/test/test.py [hunk 1]
```diff
@@ -374,87 +391,88 @@ def erc20_mint_id(self, erc20):
...
-        balance_erc20 = self.createTokenAccount(mintId)
-        print ("create account balance_erc20:", balance_erc20)
-
-        deploy_result= self.loader.deploy("erc20_ctor_uninit.hex", "erc20.bin",  mintId, balance_erc20)
+        print("\ncreate token:", token)
+        balance_erc20 = self.createTokenAccount(token)
+        print("create account balance_erc20:", balance_erc20)
+        acc2 = RandomAccount()
+        acc_client = self.createTokenAccount(token, acc2)
+        print("create account acc_client:", acc_client)
```
Comment (vasiliy-zaznobin) [inline evm_loader/ERC20/test/test.py:401]: RandomAccount is used to create the second token account

File: evm_loader/ERC20/test/test.py [hunk 2]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
+            self.assertEqual(instruction, 6)  # 6 means OnReturn
+            self.assertLess(data[1], 0xd0)  # less 0xd0 - success
+            value = data[2:]
+            ret = int.from_bytes(value, "little")
+            print('erc20_withdraw:', 'OK' if ret != 0 else 'FAIL')
+            return ret
+
+    def erc20_balance(self, erc20, erc20_code):
+        input = bytes.fromhex(
             "0370a08231" +
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:254]: please, use readable method name

File: evm_loader/ERC20/test/test.py [hunk 3]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-        result = confirm_transaction(http_client, result["result"])
+        result = send_transaction(client, trx, self.acc)
+        print(result)
         messages = result["result"]["meta"]["logMessages"]
-        res = messages[messages.index("Program log: succeed") + 1]
-        if not res.startswith("Program log: "):
-            raise Exception("Invalid program logs: no result")
+        if any(search("Program %s failed" % evm_loader_id, m) for m in messages):
+            raise Exception("Invalid program logs: Program %s failed" % evm_loader_id)
         else:
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:275]: this is unnecessary.
please , remove it

File: evm_loader/ERC20/test/test.py [hunk 4]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
+            instruction = data[0]
+            self.assertEqual(instruction, 6)  # 6 means OnReturn
+            self.assertLess(data[1], 0xd0)  # less 0xd0 - success
+            value = data[2:]
+            balance = int.from_bytes(value, "big")
+            return balance
+
+    def erc20_transfer(self, erc20, erc20_code, eth_to, amount):
         input = bytearray.fromhex(
             "a9059cbb" +
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:287]: please, use readable method name

File: evm_loader/ERC20/test/test.py [hunk 5]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
+        res = messages[-1]
+        print("res:", res)
+        if any(search("Program %s failed" % evm_loader_id, m) for m in messages):
+            raise Exception("Invalid program logs: Program %s failed" % evm_loader_id)
         else:
-            if int(res[13:], 16) == 1:
-                print("transfer OK")
-            else:
-                print("transfer Fail")
+            return res
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:320]: where is parsing ?

File: evm_loader/ERC20/test/test.py [hunk 6]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
+        result = send_transaction(client, trx, self.acc)
         messages = result["result"]["meta"]["logMessages"]
-        res = messages[messages.index("Program log: succeed") + 1]
-        if not res.startswith("Program log: "):
-            raise Exception("Invalid program logs: no result")
+        res = messages[-1]
+        print("balance_ext", res)
+        if any(search("Program %s failed" % evm_loader_id, m) for m in messages):
+            raise Exception("Invalid program logs: Program %s failed" % evm_loader_id)
         else:
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:341]: this is unnecessary.
please , remove it

File: evm_loader/ERC20/test/test.py [hunk 7]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-        result = confirm_transaction(http_client, result["result"])
+        result = send_transaction(client, trx, self.acc)
         messages = result["result"]["meta"]["logMessages"]
-        res = messages[messages.index("Program log: succeed") + 1]
-        if not res.startswith("Program log: "):
-            raise Exception("Invalid program logs: no result")
+        res = messages[-1]
+        if any(search("Program %s failed" % evm_loader_id, m) for m in messages):
+            raise Exception("Invalid program logs: Program %s failed" % evm_loader_id)
         else:
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:370]: this is unnecessary.
please , remove it

File: evm_loader/ERC20/test/test.py [hunk 8]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-            return res[13:]
+            return res
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:371]: where is parsing ?

File: evm_loader/ERC20/test/test.py [hunk 9]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-
-        self.changeOwner(balance_erc20, erc20Id)
-        print("balance_erc20 owner changed to {}".format(erc20Id))
+        token = self.createToken()
+        print("token:", token)
+
+        balance_erc20 = self.createTokenAccount(token, self.erc20Id_precalculated)
+        print("balance_erc20:", balance_erc20)
+
+        (erc20Id, erc20Id_ether, erc20_code) = self.loader.deploy_erc20("ERC20.bin"
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:380]: ERC20.bin is not exists in the repo

File: evm_loader/ERC20/test/test.py [hunk 10]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
+        assert (self.erc20Id_precalculated == erc20Id)
+
+        print("erc20 balance_ext():", self.erc20_balance_ext(erc20Id, erc20_code))
+        print("erc20 mint_id():", self.erc20_mint_id(erc20Id, erc20_code))
+
+        client_acc = self.createTokenAccount(token)
+
+        # Remove changeOwner because of createTokenAccount(token, self.erc20Id_precalculated)
+        # self.changeOwner(balance_erc20, erc20Id)
+        # print("balance_erc20 owner changed to {}".format(erc20Id))
```
Comment (sinev-valentine) [inline evm_loader/ERC20/test/test.py:398]: please, remove these comments

**Review by sinev-valentine — CHANGES_REQUESTED** [review]

File: evm_loader/ERC20/test/test.py (hunk 9 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py:380]: ERC20.bin is built from ERC20.sol every CI session

File: evm_loader/ERC20/test/test.py [hunk 11]
```diff
@@ -88,42 +27,9 @@ def call(self, arguments):
...
-
-    def ether2program(self, ether):
-        cli = SolanaCli(self.solana_url)
-        output = cli.call("create-program-address {} {}".format(ether.hex(), self.loader_id))
-        items = output.rstrip().split('  ')
-        return (items[0], int(items[1]))
-
-    def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
+
+class EvmLoaderERC20(EvmLoader):
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: EvmLoaderERC20 is derived from EvmLoader and the test use some function from EvmLoader
For example, ether2program

[the remark this answers is not in the document]
Comment (sinev-valentine) [reply evm_loader/ERC20/test/test.py:31]: Method deployERC20()  does  not need  ether2program().
The method needs the caller. You may add caller to arg list:
def deploy_erc20(loader, location_hex, location_bin, mintId, balance_erc20, caller):

File: evm_loader/ERC20/test/test.py (hunk 9 above)
Comment (sinev-valentine) [reply evm_loader/ERC20/test/test.py:375]: CI - it is next task.  Is it right?

File: evm_loader/ERC20/test/test.py (hunk 10 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 9 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py:375]: The test runs in CI with this PR

File: evm_loader/ERC20/test/test.py [hunk 12]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
 
-        info = http_client.get_account_info(cls.caller)
+        print('cls.caller_ether:', cls.caller_ether.hex())
+        (cls.caller, cls.caller_nonce) = cls.loader.ether2program(cls.caller_ether)
+        print('cls.caller:', cls.caller)
+
+        while getBalance(cls.acc.public_key()) == 0:
+            tx = client.request_airdrop(cls.acc.public_key(), 10 * 10 ** 9)
+            confirm_transaction(client, tx['result'])
+            time.sleep(1)
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py:70]: this sleep is used to wait airdrop

File: evm_loader/ERC20/test/test.py [hunk 13]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-            raise Exception("cannot get keypair path")
-        path = res[len(substr):]
-        with open(path.strip(), mode='r') as file:
-            pk = (file.read())
-            nums = list(map(int, pk.strip("[]").split(',')))
-            nums = nums[0:32]
-            values = bytes(nums)
-            cls.acc = Account(values)
-
+        cls.wallet = RandomAccount(wallet_path())
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py:55]: RandomAccount(wallet_path()) uses filesystem-wallet

File: evm_loader/ERC20/test/test.py [hunk 14]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
     from web3 import Web3
     return bytes(Web3.keccak(bytes(PublicKey(public_key)))[-20:])
 
 
 def getBalance(account):
-    return http_client.get_balance(account)['result']['value']
+    return client.get_balance(account)['result']['value']
 
 
 class EvmLoaderTests(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fix

File: evm_loader/ERC20/test/test.py [hunk 15]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-                return self.deploy(location_bin)
+                return self.deployChecked(location_bin, caller, caller_ether)
+
 
 def solana2ether(public_key):
     from web3 import Web3
     return bytes(Web3.keccak(bytes(PublicKey(public_key)))[-20:])
 
 
 def getBalance(account):
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py [hunk 16]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-            print("solana caller:", caller)
+            print("solana caller:", caller, 'cls.caller:', cls.caller)
 
-        if getBalance(cls.acc.public_key()) == 0:
-            print("Create user account...")
-            tx = http_client.request_airdrop(cls.acc.public_key(), 10*10**9)
-            confirm_transaction(http_client, tx['result'])
-            # balance = http_client.get_balance(cls.acc.public_key())['result']['value']
-            print("Done\n")
+        cls.caller_nonce = getTransactionCount(client, cls.caller)
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py:81]: Fixed

File: evm_loader/ERC20/test/test.py [hunk 17]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
+            res = spl.call("create-account {} --owner {}".format(token, owner))
         if not res.startswith("Creating account "):
-            raise Exception("create account error")
+            raise Exception("create account error %s" % res)
         else:
             return res[17:61]
 
-    def changeOwner(self, acc, owner):
+    @staticmethod
+    def changeOwner(acc, owner):
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py:116]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 8 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py:342]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 4 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 3 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 6 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 7 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py:341]: Fixed

File: Dockerfile [hunk 18]
```diff
@@ -55,5 +60,11 @@ COPY --from=spl-token-builder /opt/spl-token /opt/
 COPY --from=contracts /opt/ /opt/solidity/
 COPY evm_loader/*.py evm_loader/deploy-test.sh /opt/
 
+COPY --from=contracts /ERC20/ /opt/ERC20/
+COPY evm_loader/ERC20/test/* evm_loader/deploy-test.sh /opt/ERC20/
+RUN ln -s /opt/evm_loader.so /opt/ERC20/evm_loader.so && \
+    ln -s /opt/neon-cli /opt/ERC20/neon-cli && \
+    ln -s /opt/spl-token /opt/ERC20/spl-token
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply Dockerfile]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sinev-valentine) [reply evm_loader/ERC20/test/test.py:55]: Wallet was created by CI.
no need to create a new wallet.
The airdrop transaction takes a time

File: evm_loader/ERC20/test/test.py (hunk 5 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py [hunk 19]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-        input = bytearray.fromhex(
+            src_data = result['result']['meta']['innerInstructions'][0]['instructions'][2]['data']
+            data = base58.b58decode(src_data)
+            instruction = data[0]
+            self.assertEqual(instruction, 6)  # 6 means OnReturn
+            self.assertLess(data[1], 0xd0)  # less 0xd0 - success
+            value = data[2:]
+            ret = int.from_bytes(value, "little")
+            print('erc20_deposit:', 'OK' if ret != 0 else 'FAIL')
+            return ret
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py [hunk 20]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-
-    def erc20_withdraw(self, receiver, amount, erc20, balance_erc20, mint_id):
-        input = bytearray.fromhex(
+            src_data = result['result']['meta']['innerInstructions'][0]['instructions'][2]['data']
+            data = base58.b58decode(src_data)
+            instruction = data[0]
+            self.assertEqual(instruction, 6)  # 6 means OnReturn
+            self.assertLess(data[1], 0xd0)  # less 0xd0 - success
+            value = data[2:]
+            ret = int.from_bytes(value, "little")
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

**Review by sinev-valentine — APPROVED** [review]

File: evm_loader/ERC20/test/test.py [hunk 21]
```diff
@@ -78,13 +79,13 @@ def setUpClass(cls):
             print("Done")
             print("solana caller:", caller, 'cls.caller:', cls.caller)
 
-        cls.caller_nonce = getTransactionCount(client, cls.caller)
+        cls.trx_count = getTransactionCount(client, cls.caller)
 
         print('Account:', cls.acc.public_key(), bytes(cls.acc.public_key()).hex())
-        print("Caller:", cls.caller_ether.hex(), cls.caller_nonce, "->", cls.caller,
+        print("Caller:", cls.caller_ether.hex(), cls.trx_count, "->", cls.caller,
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py [hunk 22]
```diff
@@ -138,383 +44,418 @@ def deployERC20(self, location_hex, location_bin,  mintId, balance_erc20):
...
-        info = http_client.get_account_info(cls.caller)
+        print('cls.caller_ether:', cls.caller_ether.hex())
+        (cls.caller, cls.caller_nonce) = cls.loader.ether2program(cls.caller_ether)
+        print('cls.caller:', cls.caller)
+
+        while getBalance(cls.acc.public_key()) == 0:
+            tx = client.request_airdrop(cls.acc.public_key(), 10 * 10 ** 9)
+            confirm_transaction(client, tx['result'])
+            time.sleep(1)
+            balance = client.get_balance(cls.acc.public_key())['result']['value']
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: evm_loader/ERC20/test/test.py (hunk 8 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test.py]: Fixed

File: Dockerfile [hunk 23]
```diff
@@ -54,6 +59,9 @@ COPY --from=evm-loader-builder /opt/evm_loader/cli/target/release/neon-cli /opt/
 COPY --from=spl-token-builder /opt/spl-token /opt/
 COPY --from=contracts /opt/ /opt/solidity/
 COPY evm_loader/*.py evm_loader/deploy-test.sh /opt/
+COPY --from=contracts /ERC20/ /opt/ERC20/
+COPY evm_loader/ERC20/test/* evm_loader/deploy-test.sh /opt/ERC20/
+RUN ln -s /opt/evm_loader.so /opt/ERC20/evm_loader.so
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply Dockerfile]: Fixed. 
Move the test to run it with the others

File: .buildkite/steps/deploy-test.sh [hunk 24]
```diff
@@ -24,12 +24,20 @@ function cleanup_docker {
...
 sleep 10
 
+echo "Run ERC20 tests..."
+docker run --rm --network evm_loader-deploy_test-net -ti \
+     -e SOLANA_URL=http://solana:8899 \
+     --workdir /opt/ERC20/ \
+     ${EXTRA_ARGS:-} \
+     $EVM_LOADER_IMAGE 'deploy-test.sh'
+echo "Run ERC20 tests return"
+
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply .buildkite/steps/deploy-test.sh]: Fixed
Remove this block

**Review by s-medvedev — APPROVED** [review]

**Pull request merged**